shellscript/shellscript.go

```diff
@@ -3,13 +3,85 @@ package shellscript
 
 import (
 	"bufio"
+	"fmt"
 	"os"
 	"path/filepath"
 	"strings"
 
 	"github.com/buildkite/shellwords"
+	"golang.org/x/exp/slices"
 )
 
+// Magic numbers for various types of executable files
+var (
+	// Linux and many other Unix-like systems use ELF binaries
+	ELFMagic = []byte{0x7F, 0x45, 0x4C, 0x46}
+
+	// Windows uses MS-DOS MZ-style executables. This includes PE32 and PE32+ (which is 64-bit 🙃) binaries,
+	// which a are what modern windowses tend to use.
+	MZMagic = []byte{0x4D, 0x5A}
+
+	// macOS uses Mach-O binaries. There are two variants: fat and thin.
+	// Thin binaries are further divided into 32-bit and 64-bit variants,
+	// which are furtherer divided by processor endianess. It's a mess.
+
+	// For "fat" binaries, which contain multiple architectures
+	MachOFatMagic = []byte{0xCA, 0xFE, 0xBA, 0xBE}
+
+	// For "thin" binaries, which contain a single architecture
+	// (ie 32 and 64 bit variants, on top of different processor architectures)
+	MachOThin32Magic = []byte{0xFE, 0xED, 0xFA, 0xCE}
+	MachOThin64Magic = []byte{0xFE, 0xED, 0xFA, 0xCF}
+
+	// Mach-O thin binaries can also be in reverse byte order, apparently?
+	// I think this is for big endian processors like PowerPC, in which case we don't need these here,
+	// but we might as well include them for completeness.
+	MachOThin32ReverseMagic = []byte{0xCE, 0xFA, 0xED, 0xFE}
+	MachOThin64ReverseMagic = []byte{0xCF, 0xFA, 0xED, 0xFE}
+
+	BinaryMagicks = [][]byte{
+		ELFMagic,
+		MZMagic,
+		MachOFatMagic,
+		MachOThin32Magic,
+		MachOThin64Magic,
+		MachOThin32ReverseMagic,
+		MachOThin64ReverseMagic,
+	}
+)
+
+// ExecutableType returns the type of file executable at the given path.
+// The file at the given path is assumed to be an executable, and executability is not checked.
+func IsBinaryExecutable(path string) (bool, error) {
+	f, err := os.Open(path)
+	if err != nil {
+		return false, fmt.Errorf("open file %q: %w", path, err)
+	}
+
+	defer f.Close()
+	r := bufio.NewReader(f)
+	magic, err := r.Peek(4)
```

Comment:
I was surprised that `Peek` was the only method that you could ask for a specific number of bytes and get them as a slice… I assumed there's be a `Read…` method for that, which would have made slightly more sense in this context given we don't really care about not advancing the reader, but… `Peek` it is!

Reply:
you can also do:
```Go
b := make([]byte, 4)
err := f.Read(b)
```
which will read at most 4 bytes into b, as will
```Go
bytes, err := ioutil.ReadAll(io.LimitReader(f, 4))
```

perhaps either of those is clearer? i'm no 100% sure

Reply:
Ooh fancy LimitReader composition 👌🏼

I like `r.Peek(4)` better though, because it's less words and still quite clear 👍🏼